htsget-search/src/storage/url/mod.rs

```diff
@@ -0,0 +1,1178 @@
+              ),
+          ),
+          Url::new("http://example.com/htsnexus_test_NA12878.bam.c4gh").with_headers(
+            Headers::default()
+              .with_header("authorization", "secret")
+              .with_header(CLIENT_PUBLIC_KEY_NAME, &public_key)
+              .with_header("Range", format!("bytes={}-{}", 124 + 2556996, 2598043 - 1)),
+          ),
+        ],
+      );
+
+      assert_eq!(response, expected_response);
+
+      let (bytes, _) = test_bam_file_byte_ranges(
+        response,
+        default_dir().join("data/crypt4gh/htsnexus_test_NA12878.bam.c4gh"),
+      )
+      .await;
+
+      let (expected_bytes, _) = test_bam_file_byte_ranges(
+        expected_bam_response(),
+        default_dir().join("data/bam/htsnexus_test_NA12878.bam"),
+      )
+      .await;
+
+      test_bam_crypt4gh_byte_ranges(bytes.clone(), expected_bytes).await;
+    })
+    .await;
+  }
+
+  fn expected_bam_response() -> HtsgetResponse {
+    HtsgetResponse::new(
+      Format::Bam,
+      vec![
+        Url::new("http://example.com/htsnexus_test_NA12878.bam")
+          .with_headers(
+            Headers::default()
+              .with_header("authorization", "secret")
+              .with_header("Range", "bytes=0-4667"),
+          )
+          .with_class(Header),
+        Url::new("http://example.com/htsnexus_test_NA12878.bam")
+          .with_headers(
+            Headers::default()
+              .with_header("authorization", "secret")
+              .with_header("Range", "bytes=256721-1065951"),
+          )
+          .with_class(Body),
+        Url::new("http://example.com/htsnexus_test_NA12878.bam")
+          .with_headers(
+            Headers::default()
+              .with_header("authorization", "secret")
+              .with_header("Range", "bytes=2596771-2596798"),
+          )
+          .with_class(Body),
+      ],
+    )
+  }
+
+  fn test_client() -> Client<HttpsConnector<HttpConnector>> {
+    Client::builder().build(
+      HttpsConnectorBuilder::new()
+        .with_native_roots()
+        .https_or_http()
+        .enable_http1()
+        .enable_http2()
+        .build(),
+    )
+  }
+
+  pub(crate) async fn with_url_test_server<F, Fut>(test: F)
+  where
+    F: FnOnce(String) -> Fut,
+    Fut: Future<Output = ()>,
+  {
+    let (_, base_path) = create_local_test_files().await;
+    with_test_server(base_path.path(), test).await;
+  }
+
+  async fn test_auth<B>(
+    request: Request<B>,
+    next: Next<B>,
+  ) -> result::Result<Response, StatusCode> {
+    let auth_header = request
+      .headers()
+      .get(AUTHORIZATION)
+      .and_then(|header| header.to_str().ok());
+
+    match auth_header {
+      Some("secret") => Ok(next.run(request).await),
+      _ => Err(StatusCode::UNAUTHORIZED),
+    }
+  }
+
+  pub(crate) async fn with_test_server<F, Fut>(server_base_path: &Path, test: F)
+  where
+    F: FnOnce(String) -> Fut,
+    Fut: Future<Output = ()>,
+  {
```

Comment:
This is the mock `UrlStorage` server used by the test here: https://github.com/umccr/htsget-rs/pull/223/files#r1452880329